File: 0132-palindrome-partitioning-ii/0132-palindrome-partitioning-ii.py
class Solution(object):
    def minCut(self, s):
        """
        :type s: str
        :rtype: int
        """
        n = len(s)
        if n <= 1:
            return 0

        # pal[i][j] = True if s[i:j+1] is a palindrome
        pal = [[False] * n for _ in range(n)]

        # Single characters
        for i in range(n):
            pal[i][i] = True

        # Length >= 2 substrings
        for length in range(2, n + 1):
            for i in range(n - length + 1):
                j = i + length - 1
                if s[i] == s[j]:
                    if length == 2:
                        pal[i][j] = True
                    else:
                        pal[i][j] = pal[i + 1][j - 1]

        # cuts[i] = min cuts for s[:i+1]
        cuts = [0] * n
        for i in range(n):
            if pal[0][i]:
                cuts[i] = 0
            else:
                # worst case: cut before every character
                min_cut = i
                for j in range(1, i + 1):
                    if pal[j][i]:
                        min_cut = min(min_cut, cuts[j - 1] + 1)
                cuts[i] = min_cut

        return cuts[n - 1]

File: 0132-palindrome-partitioning-ii/0132-palindrome-partitioning-ii.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int minCut(string s) {
        int n = (int)s.size();
        vector<vector<char>> pal(n, vector<char>(n, 0));

        // pal[i][j] for all i<=j
        for (int i = n - 1; i >= 0; --i) {
            for (int j = i; j < n; ++j) {
                if (s[i] == s[j] && (j - i <= 2 || pal[i + 1][j - 1])) {
                    pal[i][j] = 1;
                }
            }
        }

        vector<int> dp(n, INT_MAX);

        for (int i = 0; i < n; ++i) {
            if (pal[0][i]) {
                dp[i] = 0;
            } else {
                for (int j = 1; j <= i; ++j) {
                    if (pal[j][i]) {
                        dp[i] = min(dp[i], dp[j - 1] + 1);
                    }
                }
            }
        }
        return dp[n - 1];
    }
};
